torch/csrc/autograd/engine.cpp

```diff
@@ -726,10 +732,17 @@ auto Engine::execute(const edge_list& roots,
   // Lock post_callbacks_lock_ before clearing final_callbacks_
   ClearCallbacks _cb_guard(final_callbacks_, post_callbacks_lock_);
 
+  // A frech first time Engine::execute call should start on the CPU device, initialize
+  // a new thread local ready queue on CPU or reuse the existing one (if there is one
+  // allocated already), then memorize the local_ready_queue in GraphTask
```

Comment:
You don't want to reuse it right? As it might not be empty in the async case?
You should clear the thread local ready queue before exiting this right?

Reply:
in the async case, it is using `graph_task->cpu_ready_queue_` instead of local_ready_queue, and for reentrant calls in the async case, we are creating a new local ready queue and execute the reentrant call with that ready queue. 

we can clear the `local_ready_queue` before we exit `Engine::execute`, but i think that might not needed as we either 1: successfully finish all nodetasks, local_ready_queue already empty 2. error thrown to the user, which they need to restart their program and local_ready_queue will be recreated.

Reply:
> error thrown to the user, which they need to restart their program and local_ready_queue will be recreated.

You can catch the error in python and keep working.
So you would leak extra tasks that the new thread will have to do before doing its own work.

Reply:
yeah that make sense. I don't know if user would like to just catch the error and continue the execution, but that decision probably should be put on the user side. If that's the valid use case, I will clear the local ready queue before engine::execute exit.

Reply:
Regarding tasks being left behind in the ready queue, isn't the behavior same in this PR compared to what we have today? IIUC, if evaluate_function throws an exception we save the exception but continue executing the rest of the graph (unless exit_on_error_ is set to true, which is only set in the distributed autograd case). We would probably have similar behavior here as well?

Now, if something else throws an exception that we didn't catch in the engine itself, that might actually be a bug in the engine itself that needs to be fixed.

Reply:
Yeah agree with @pritamdamania I think this does not change the behavior compare to the current autograd enigne. if it's not a async case, we might not need to clear the local_ready_queue to preserve the old behavior.